src/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -32,7 +33,7 @@ internal struct WriteStackFrame
         // The current property.
         public bool PropertyEnumeratorActive;
         public ExtensionDataWriteStatus ExtensionDataStatus;
-        public IEnumerator PropertyEnumerator;
+        public Dictionary<string, JsonPropertyInfo>.Enumerator PropertyEnumerator;
```

Comment:
To understand this a bit better.

PropertyCache is a `Dictionary<string, JsonPropertyInfo>`
https://github.com/dotnet/corefx/blob/7829d4acc8303a42e222415c7eb829fc3f8264cc/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleObject.cs#L29

And by saving it as an `IEnumerator` field, we were unnecessarily boxing `Dictionary<string, JsonPropertyInfo>.Enumerator` as an `IEnumerator`

And then when using this `IEnumerator`, casting back to `(KeyValuePair<string, JsonPropertyInfo>)` incurred additional allocations.

Is that correct?

Reply:
The 2nd box is casting `KeyValuePair<string, JsonPropertyInfo>` to an `object` when implementing `IEnumerator.Current`.

Reply:
Cory is correct; it's not that the cast to `KeyValuePair<string, JsonPropertyInfo>` boxed... rather, that unboxed.  It's that because we were storing the enumerator as the non-generic `IEnumerator`, which has an `object Current { get }` property, every time `Current` was accessed, it was boxing the `KeyValuePair<string, JsonPropertyInfo>`.